calyx-py/test/correctness/fifo.py

```diff
@@ -247,57 +290,93 @@ def add_main(prog, fifo):
+    # It will run until the `err` flag is raised by the `fifo` component or
+    # until it reaches the end of the command list.
+
+    i = main.reg("i", 32)  # The index of the command we're currently processing
+    j = main.reg("j", 32)  # The index on the answer-list we'll write to
+    command = main.reg("command", 32)  # The command we're currently processing
+
+    zero_i = set_flag_reg(main, i, 0, "zero_i")  # zero out `i`
+    zero_j = set_flag_reg(main, j, 0, "zero_j")  # zero out `j`
+    incr_i = add_incr(main, i, "add3", "incr_i")  # i = i + 1
+    incr_j = add_incr(main, j, "add4", "incr_j")  # j = j + 1
+    err_eq_zero = add_eq_mem(main, err, 0, 0, "err_eq_0", 1)  # is the `err` flag down?
+    # i_eq_15 = add_eq(main, i.out, 15, "i_eq_15", 32)  # is `i` == 15?
+    # raise_err = mem_store(main, err, 0, cb.const(1, 1), "raise_err")  # set `err` to 1
+    read_command = mem_load_to_reg(main, commands, i.out, command, "read_command")
+    command_eq_zero = add_eq(main, command.out, 0, "command_eq_zero", 32)
+    write_ans = mem_store_from_mem(main, ans_mem, j.out, ans, 0, "write_ans")
+
+    main.control += [
+        zero_i,
+        zero_j,
+        cb.while_(
+            err_eq_zero[0].out,
+            err_eq_zero[1],  # Run while the `err` flag is down
+            [
+                read_command,  # Read the command at `i`
+                cb.if_(
+                    # Is this a pop or a push?
+                    command_eq_zero[0].out,
+                    command_eq_zero[1],
+                    [  # a pop
+                        cb.invoke(  # first we call pop
+                            fifo,
+                            in_pop=cb.const(1, 1),
+                            in_push=cb.const(1, 0),
+                            ref_ans=ans,
+                            ref_err=err,
+                        ),
+                        cb.if_(  # but was it successful?
+                            err_eq_zero[0].out,
+                            err_eq_zero[1],
+                            [  # yes, it was successful: write the answer to ans_mem
+                                write_ans,
+                                incr_j,
+                            ],
+                        ),
+                        # AM:
+                        # The if-check above is giving me trouble:
+                        # I need to check on the `err` flag, but I am getting complaints
+                        # from Calyx because the same `err` flag is checked in the while-loop.
+                    ],
+                    cb.invoke(  # a push
+                        fifo,
+                        in_pop=cb.const(1, 0),
+                        in_push=cb.const(1, 1),
+                        in_payload=command.out,
+                        ref_ans=ans,
+                        ref_err=err,
+                    ),
+                ),
+                incr_i,  # Increment the command index
+                # cb.if_(
```

Comment:
These lines are commented out for a similar reason. Same issue: I want to access a memory that is already of interest to the while loop. In this case I want to write to it.

Reply:
Are you using `std_mem` or `seq_mem`. If you're using `std_mem`, I recommend switching to `seq_mem` since we'll deprecate the former soon (#1261) because they are not easily mapped onto good physical designs.